File: Python/threeSum.py
#!/usr/bin/env python3
# encoding: utf-8

"""
@Filename: threeSum.py
@Function: 三数之和
@Link: https://leetcode-cn.com/problems/3sum/
@Python Version: 3.8
@Author: Wei Li
@Date：2021-08-10
"""

from typing import List

# 排序 + 双指针策略 + 去重策略
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        num_list = len(nums)
        nums.sort()
        ans = list()

        for first_idx in range(num_list):
            if first_idx > 0 and nums[first_idx] == nums[first_idx - 1]:
                continue

            third_idx = num_list - 1
            target = -nums[first_idx]

            for second_idx in range(first_idx + 1, num_list):
                if second_idx > first_idx + 1 and nums[second_idx] == nums[second_idx - 1]:
                    continue

                while second_idx < third_idx and nums[second_idx] + nums[third_idx] > target:
                    third_idx -= 1

                if second_idx == third_idx:
                    break

                if nums[second_idx] + nums[third_idx] == target:
                    ans.append([nums[first_idx], nums[second_idx], nums[third_idx]])

        return ans


# -------------------------
if __name__ == "__main__":
    nums = [-1, 0, 1, 2, -1, -4]

    solution = Solution()
    three_sum_idx = solution.threeSum(nums)
    print(f"The solution of this problem is : {three_sum_idx}")

File: C++/threeSum.cpp
/**
 * @File    : threeSum.cpp
 * @Brief   : 三数之和
 * @Link    : https://leetcode-cn.com/problems/3sum/
 * @Author  : Wei Li
 * @Date    : 2021-08-10
*/

#include <iostream>
#include <vector>
#include <algorithm>

class Solution
{
public:
    std::vector<std::vector<int>> threeSum(std::vector<int> &nums)
    {
        size_t num_array = nums.size();
        std::sort(nums.begin(), nums.end());
        std::vector<std::vector<int>> ans;

        // 枚举 第一个数字 a
        for (size_t first_number_idx = 0; first_number_idx < num_array; ++first_number_idx)
        {
            // 需要和上一次枚举的数字不相同(去重复)
            if (first_number_idx > 0 && nums[first_number_idx] == nums[first_number_idx - 1])
            {
                continue;
            }

            // 第三数字 c 对应的指针初始指向数组的最右端
            size_t third_number_idx = num_array - 1;
            // 利用双指针策略计算符合目标值的两个索引
            int target = -nums[first_number_idx];

            // 枚举 第二个数字 b
            for (size_t second_number_idx = first_number_idx + 1; second_number_idx < num_array; ++second_number_idx)
            {
                // 需要和上一次枚举的数字不相同(去重复)
                if (second_number_idx > first_number_idx + 1 && nums[second_number_idx] == nums[second_number_idx - 1])
                {
                    continue;
                }

                // 需要保证 b 的指针在 c 的指针的左侧 双指针策略
                while (second_number_idx < third_number_idx && nums[second_number_idx] + nums[third_number_idx] > target)
                {
                    --third_number_idx;
                }

                // 如果指针重合，随着 b 后续的增加
                // 就不会有满足 a+b+c=0 并且 b<c 的 c 了，可以退出循环
                if (second_number_idx == third_number_idx)
                {
                    break;
                }

                if (nums[second_number_idx] + nums[third_number_idx] == target)
                {
                    ans.push_back({nums[first_number_idx], nums[second_number_idx], nums[third_number_idx]});
                }
            }
        }
        return ans;
    }
};

template <typename T>
void DispalyElement(const T &container)
{
    for (auto element : container)
    {
        for (auto index : element)
        {
            std::cout << index << " ";
        }
        std::cout << std::endl;
    }
    std::cout << std::endl;
}

// --------------------------------
int main(int argc, char **argv)
{
    std::vector<int> nums{-1, 0, 1, 2, -1, -4};

    Solution solution;
    auto three_sum_idx = solution.threeSum(nums);
    DispalyElement(three_sum_idx);

    return 0;
}
